### DISCUSSION TITLE
set.filter behavior/performance

### DISCUSSION THREAD

**Discussion by cgjohannsen** [cvc5/cvc5#10837] [Q&A] [answered]
Hi! 

I have a question regarding the theory of sets as it's implemented in CVC5, especially the sets.filter function. The example I'm working off of defines a set of integers and a predicate that returns true if the given integer is between 0 and 1. 
```
(set-logic HO_ALL)
(set-option :full-saturate-quant true)

(declare-fun IntSet () (Set Int))
(define-fun filter_eq_zero_one ((i Int)) Bool (and (>= i 0) (<= i 1)))
```
The property I want to check is that every integer that is a member of this set is between 0 and 1. Using a filter, I define this property as:
```
(assert 
    (forall ((i Int)) (=> (set.member i (set.filter filter_eq_zero_one IntSet)) (and (>= i 0) (<= i 1))))
)
```
for which CVC5 outputs `unknown`. But if instead I apply the check as a condition on `i` as in
```
(assert 
    (forall ((i Int)) (=> (and (set.member i IntSet) (filter_eq_zero_one i)) (and (>= i 0) (<= i 1))))
)
```
CVC5 outputs `sat`. Is there any intuition as to why the set.filter version would struggle as opposed to the conditional check on `i`? Is this a deficiency when using quantifiers in the theory of sets? I have the two complete files attached for reference.
[simplified_0.1.txt](https://github.com/user-attachments/files/15514418/simplified_0.1.txt)
[simplified_0.2.txt](https://github.com/user-attachments/files/15514420/simplified_0.2.txt)

Thanks for any help/guidance!

**Comment by mudathirmahgoub** [comment, marked as answer]
Setting the option below would make the example with `set.filter` return true. 
```smt2
(set-option :mbqi true)
``` 
By the way, the quantified formula without `set.filter` is basically a tautology, so `cvc5` answers sat immediately .
